include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -12,59 +13,116 @@
+    unsigned char* base;
+    unsigned char* top;
 
     block(
-        std::size_t size_,
-        block* next_)
-        : size(size_)
-        , top(reinterpret_cast<
-            std::uintptr_t>(this+1))
-        , next(next_)
-    {
-    }
-
-    void*
-    alloc(
+        unsigned char* data,
         std::size_t n,
-        std::size_t align) noexcept
+        block* head)
+        : size(n)
+        , next(head)
+        , base(data)
+        , top(data)
     {
-        // must be power of 2
-        BOOST_ASSERT(
-            (align & (align - 1)) == 0);
-        auto i = top;
-        if((i & (align - 1)) != 0)
-            i = (i | (align - 1)) + 1;
-        if(i + n > (reinterpret_cast<
-            std::uintptr_t>(this+1) + size))
-            return nullptr;
-        top = i + n;
-        return reinterpret_cast<void*>(i);
     }
 };
 
+template<typename Block>
+void*
+monotonic_resource::
+allocate_in_block(
+    Block& blk,
+    std::size_t size,
+    std::size_t align)
+{
+    // must be power of 2
+    BOOST_ASSERT((align & (align - 1)) == 0);
+    std::size_t block_size = 
+        blk.size - (blk.top - blk.base);
+    void* top = blk.top;
+    // KRYSTIAN TODO: std::align isnt very compatible
+    // with our block design; maybe change this
+    const auto aligned = 
+        std::align(align, size, top, block_size);
+    if (aligned)
+        blk.top = static_cast<unsigned char*>(aligned) + size;
+    return aligned;
+}
+
 auto
 monotonic_resource::
-alloc_block(std::size_t size) ->
+allocate_new_block(std::size_t size) ->
     block&
+{  
+    /*-------------------------------------  
+             block storage layout:        
+           [ n ] = n allocated bytes
+      
+                header         data
+           [ sizeof(block) ] [ size ]
+      
+        data is guaranteed to be aligned
+          to alignof(std::max_align_t)
+    -------------------------------------*/
+    const auto bytes = size + sizeof(block);
+    const auto data = static_cast<unsigned char*>(
+        ::operator new(bytes));
+    head_ = ::new(static_cast<void*>(data))
+        block(data + sizeof(block), size, head_);
+    return *head_;
+}
+
+std::size_t
+monotonic_resource::
+next_block_size(
+    std::size_t requested,
+    std::true_type)
 {
-    auto const n = (
-        size + sizeof(block) - 1) /
-        sizeof(block);
-    auto const bytes =
-        (n + 1) * sizeof(block);
-    auto& b = *::new(
-        ::operator new(bytes)) block(
-            n * sizeof(block), head_);
-    head_ = &b;
-    return b;
+    // returns the closest power of two
+    // that is greater than requested
+    if(requested >= max_block_size_ >> 1)
+        return max_block_size_;
+    if(requested <= min_block_size_)
+        return min_block_size_;
+    std::size_t next = 1;
+    for(; requested; requested >>= 1, next <<= 1);
+    return next;
+}
+
+std::size_t
+monotonic_resource::
+next_block_size(
+    std::size_t requested,
+    std::false_type)
```

Comment:
What the hell is this true_type false_type? I have no idea. Couldn't we just name the functions differently?